src/Umbraco.Core/Cache/IDataTypeReadCache.cs

```diff
@@ -0,0 +1,8 @@
+using Umbraco.Cms.Core.Models;
+
+namespace Umbraco.Cms.Core.Cache;
+
+public interface IDataTypeReadCache
+{
+    IDataType? GetDataType(int id);
```

Comment:
Adding guid / udi overloads make sense, but than we should propably cache them as IEntityService is expensive to talk with. This pr fix it by introducing intermediate cache, so adding calll to IEntityService  would not make sense. 
I guess we can add to memory cache list of guid->id which would read from same cache as int id?

Reply:
I think we should _only_ have methods that accepts GUID (key) here. We are (sloooowly) moving away from integer IDs in public APIs, so adding a new interface with integer IDs seems counter productive to that end.

I have been through the interface usings, and it seems to me that there is a `DataTypeKey` available (instead of `DataTypeId`) whereever it is used.

Reply:
While looking further into this issue, I actually noticed that the `GetDataType(int)` uses the repository cache, but the `GetDataType(Guid)` does a query each time... Besides the query overhead, this causes it to return a new `DataType` instance each time, which needs to initialize (deserialize) the configuration on the first access, hence the performance issues!

Reply:
@ronaldbarendse even with repository cache getting by guid was causing performance issue, so the fix would be still valid :). The issue is related to reuse of locks by umbraco and locking cache quite a lot, so even so we use cache repository in most of places -> it is waiting for locks so it takes a long time

Reply:
I actually don't think the issue is related to locks, but more due to the fact it's doing a deep clone and/or tracks dirty properties each time it's retrieved from the cache:

https://github.com/umbraco/Umbraco-CMS/blob/60a5b19dc9fe24f5379a253e1f3fdcb9cb537dfc/src/Umbraco.Core/Cache/AppCaches.cs#L72-L76

Besides this, the fact that the `GetDataType(Guid)` didn't even use the repository cache, but did a database query each time and returned a new `DataType` instance that needed to initialize/deserialize the configuration, were the biggest factors for the performance issues... The new `IDataTypeConfigurationCache` added in this PR should fix all these issues though! 🤞🏻